pkg/testutils/identity/allocator.go

```diff
@@ -20,66 +20,153 @@ func (i *IdentityAllocatorOwnerMock) GetNodeSuffix() string {
+		ID:     identity.NumericIdentity(id),
+		Labels: lbls,
+	}
+	f.idToIdentity[id] = realID
+	f.idRefCount[id]++
+
+	return realID, true, nil
 }
 
-// Release does nothing.
-func (f *FakeIdentityAllocator) Release(context.Context, *identity.Identity) (released bool, err error) {
+// Release releases a fake identity. It is meant to generally mock the
+// canonical identity release logic.
+func (f *MockIdentityAllocator) Release(_ context.Context, id *identity.Identity) (released bool, err error) {
+	count, ok := f.idRefCount[int(id.ID)]
+	if !ok {
+		return false, nil
+	}
+	if count == 1 {
+		delete(f.idRefCount, int(id.ID))
+	} else {
+		count--
+		f.idRefCount[int(id.ID)] = count
+		return false, nil
+	}
+	delete(f.idToIdentity, int(id.ID))
+	delete(f.IdentityCache, id.ID)
 	return true, nil
 }
 
-// ReleaseSlice does nothing.
-func (f *FakeIdentityAllocator) ReleaseSlice(context.Context, cache.IdentityAllocatorOwner, []*identity.Identity) error {
+// ReleaseSlice wraps Release for slices.
+func (f *MockIdentityAllocator) ReleaseSlice(ctx context.Context, _ cache.IdentityAllocatorOwner, identities []*identity.Identity) error {
+	for _, id := range identities {
+		if _, err := f.Release(ctx, id); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
-// LookupIdentity is a no-op.
-func (f *FakeIdentityAllocator) LookupIdentity(ctx context.Context, lbls labels.Labels) *identity.Identity {
-	return nil
+// LookupIdentity looks up the labels in the mock identity store.
+func (f *MockIdentityAllocator) LookupIdentity(ctx context.Context, lbls labels.Labels) *identity.Identity {
+	if reservedIdentity := identity.LookupReservedIdentityByLabels(lbls); reservedIdentity != nil {
+		return reservedIdentity
+	}
+	return f.idToIdentity[f.labelsToIdentity[lbls.String()]]
 }
 
 // LookupIdentityByID returns the identity corresponding to the id if the
 // identity is a reserved identity. Otherwise, returns nil.
-func (f *FakeIdentityAllocator) LookupIdentityByID(ctx context.Context, id identity.NumericIdentity) *identity.Identity {
+func (f *MockIdentityAllocator) LookupIdentityByID(ctx context.Context, id identity.NumericIdentity) *identity.Identity {
 	if identity := identity.LookupReservedIdentity(id); identity != nil {
 		return identity
 	}
-	return nil
+	return f.idToIdentity[int(id)]
 }
 
-func (f *FakeIdentityAllocator) AllocateCIDRsForIPs(IPs []net.IP, newlyAllocatedIdentities map[string]*identity.Identity) ([]*identity.Identity, error) {
-	return nil, nil
+// AllocateCIDRsForIPs allocates CIDR identities for the given IPs. It is meant
+// to generally mock the CIDR identity allocator logic.
+func (f *MockIdentityAllocator) AllocateCIDRsForIPs(IPs []net.IP, _ map[string]*identity.Identity) ([]*identity.Identity, error) {
+	result := make([]*identity.Identity, 0, len(IPs))
+	for _, ip := range IPs {
+		id, ok := f.ipToIdentity[ip.String()]
+		if !ok {
+			id = f.localID
+			f.ipToIdentity[ip.String()] = id
+			f.idRefCount[id]++
```

Comment:
This testutils refcount increase should happen unconditionally rather than only if the IP is getting a new identity.

Reply:
^^ Using `pkg/counter` could also help avoid issues like this.